clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -238,7 +238,7 @@ public RecordsInfo info() {
         }
     }
 
-    public void setProducerState(long producerId, short producerEpoch, int baseSequence) {
+    public void setProducerState(long producerId, short producerEpoch, int baseSequence, boolean isTransactional) {
```

Comment:
To clarify, the reason I had to change this is that we need to close the record builder in order to split it. At that point, we don't have a producerId yet, so if `isTransactional` is set to true, then `MemoryRecordsBuilder.close()` will raise an exception. To get around that, this change ensures that we always have a producerId when we set `isTransactional`.